examples/basic_ranking.py

```diff
@@ -0,0 +1,217 @@
+"""
+# Recommending movies: ranking
+
+Recommender systems are often composed of two stages:
+
+1. The retrieval stage is responsible for selecting an initial set of hundreds
+   of candidates from all possible candidates. The main objective of this model
+   is to efficiently weed out all candidates that the user is not interested in.
+   Because the retrieval model may be dealing with millions of candidates, it
+   has to be computationally efficient.
+2. The ranking stage takes the outputs of the retrieval model and fine-tunes
+   them to select the best possible handful of recommendations. Its task is to
+   narrow down the set of items the user may be interested in to a shortlist of
+   likely candidates.
+
+In this tutorial, we're going to focus on the first stage, retrieval. If you are
+interested in the ranking stage, have a look at our
+[retrieval](https://github.com/keras-team/keras-rs/blob/main/examples/basic_retrieval.py)
+tutorial.
+
+In this tutorial, we're going to:
+
+1. Get our data and split it into a training and test set.
+2. Implement a ranking model.
+3. Fit and evaluate it.
+4. Test running predictions with the model.
+"""
+
+import keras
+
+# Needed for the dataset
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+"""
+## Preparing the dataset
+
+We're going to use the same data as the
+[retrieval](https://github.com/keras-team/keras-rs/blob/main/examples/basic_retrieval.py)
+tutorial. The ratings are the objectives we are trying to predict.
+"""
+
+# Ratings data.
+ratings = tfds.load("movielens/100k-ratings", split="train")
+# Features of all the available movies.
+movies = tfds.load("movielens/100k-movies", split="train")
+
+"""
+User ids are integers (represented as strings) between 1 and 943. We'll use the
+user id as an index in our model, in particular to lookup the user embedding
+from the user embedding table. So we need do know the number of users.
+"""
+
+users_count = (
+    ratings.map(lambda x: tf.strings.to_number(x["user_id"], out_type=tf.int32))
+    .reduce(tf.constant(0, tf.int32), tf.maximum)
+    .numpy()
+)
+
+"""
+Movie ids are integers (represented as strings) between 1 and 1682. We'll use
+the movie id as an index in our model, in particular to lookup the movie
+embedding from the movie embedding table. So we need do know the number of
+movies.
+"""
+
+movies_count = movies.cardinality().numpy()
+
+"""
+The inputs to the model are the user ids and movie ids and the labels are the
+ratings.
+"""
+
+
+def preprocess_rating(x):
+    return (
+        # Inputs are user ids and movie ids
+        {
+            "user_id": tf.strings.to_number(x["user_id"], out_type=tf.int32),
```

Comment:
Are these IDs guaranteed to be sequential? If they're just random int-like, we would want to reindex them from 0 to num_users

Reply:
I cheated because I know the dataset and they are sequential.

The original example was more general because it wasn't making any assumption on the IDs,  not even that they were numbers, because it's using a `StringLookup` layer as part of the model. `StringLookup` is not multi-backend so we can't do that as part of the model, it would have to be part of preprocessing. But then it gets more complicated to demonstrate inference (in the retrieval example), because the reverse lookup is needed, so we would need:
- model.predict
- convert to Tensorflow
- reverse lookup
- convert to Python
- lookup movie title

We have 3 options:
- make it fully generic and use `StringLookup` outside of the model
- assume IDs are ints and use an int to int mapping
- leave it as is (assume sequential IDs)

Reply:
> leave it as is (assume sequential IDs)

That's ok if you clearly document this gotcha

Reply:
Added some detailed language about this shortcut.